pkg/networkmanager/firewall-client.es6

```diff
@@ -0,0 +1,175 @@
+/* jshint esversion: 6 */
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2018 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import cockpit from 'cockpit';
+import service from 'service';
+
+var firewall = {
+    installed: true,
+    enabled: false,
+    services: {},
+    enabledServices: new Set()
+};
+
+cockpit.event_target(firewall);
+
+var firewalld_service = service.proxy('firewalld');
+var firewalld_dbus = cockpit.dbus('org.fedoraproject.FirewallD1', { superuser: 'try' });
+
+firewalld_service.addEventListener('changed', function () {
+    var installed = !!firewalld_service.exists;
+
+    if (firewall.installed == installed)
+        return;
+
+    firewall.installed = installed;
+    firewall.dispatchEvent('changed');
+});
+
+function fetchServiceInfos(services) {
+    return cockpit.all(services.map(service => {
+        if (firewall.services[service])
+            return firewall.services[service];
+
+        return firewalld_dbus.call('/org/fedoraproject/FirewallD1',
+                                   'org.fedoraproject.FirewallD1',
+                                   'getServiceSettings', [service])
+            .then(reply => {
+                var [ , name, description, ports ] = reply[0];
+
+                var info = {
+                    id: service,
+                    name: name,
+                    description: description,
+                    ports: ports.map(p => ({ port: p[0], protocol: p[1] }))
+                };
+
+                firewall.services[service] = info;
+                return info;
+            });
+    }))
+
+    /*
+     * Work around `cockpit.all()` returning results in individual arguments -
+     * that's just confusing.
+     */
+    .then(function () {
+        return Array.prototype.slice.call(arguments);
```

Comment:
Yeah, it's the most succinct way to do so. I can't use a fat-arrow function here, because that doesn't have the `arguments` object (took me a while to figure out). This looks confusing, because the real culprit here is `cockpit.all()` not returning a single list. I've updated the comment to be more clear.